main.go

```diff
@@ -1,9 +1,482 @@
+
+	slice := make([]float64, 0)
+	for i := 0; i < len(dmb); i++ {
+		val := math.Sqrt(math.Pow(dmb[i], 2) + math.Pow(alpha*dstretch[i], 2) + math.Pow(beta*dcolour[i], 2))
+		slice = append(slice, val)
+	}
+
+	return slice
+}
+
+
+/*
+######
+Functions for diagrams : plots, fit, etc
+######
+*/
+
+func points(x_values, y_values []float64) plotter.XYs {
+
+	// creates a plotter.XYs (cloud of points) from given 2D coordinates
+	// inputs : x_values, y_values = slices containing the x and y values respectively
+
+	pts := make(plotter.XYs, len(x_values))
+	for i := range pts {
+		pts[i].X = x_values[i]
+		pts[i].Y = y_values[i]
+	}
+
+	return pts
+}
+
+func diag(points plotter.XYs) *plot.Plot {
+
+	// plot the given points
+	// input : points = plotter.XYs created by the points function
+
+	p, err := plot.New()
+	if err != nil {
+		log.Panic(err)
+	}
+
+	p.Title.Text = "Hubble diagram"
+	p.X.Label.Text = "Redshift z"
+	p.Y.Label.Text = "Distance mu"
+	p.X.Scale = plot.LogScale{}
+	p.X.Tick.Marker = plot.LogTicks{}
+	p.Add(plotter.NewGrid())
+
+	s, err := plotter.NewScatter(points)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	s.GlyphStyle.Color = color.RGBA{R: 255, B: 0, A: 255}
+	s.GlyphStyle.Radius = vg.Points(3)
+	s.GlyphStyle.Shape = draw.PlusGlyph{}
+
+	p.Add(s)
+
+	return p
+}
+
+func modified_integral(z, omega float64) float64 {
+
+	// modified integral function where omgM is a variablef that can be optimized
+	// inputs : z = redshift, omega = omgM as a free parameter
+
+	var inte float64
+	abs := make([]float64, 0)
+	ord := make([]float64, 0)
+
+	var i float64
+	for i = 0; i <= 1000; i++ {
+		val := z * i / 1000
+		abs = append(abs, val)
+	}
+
+	for j := 0; j <= 1000; j++ {
+		val := 1 / math.Sqrt((1+abs[j]*omega)*(1+abs[j])*(1+abs[j])-(1-omega)*(2+abs[j])*abs[j])
+		ord = append(ord, val)
+	}
+
+	inte = integrate.Trapezoidal(abs, ord)
+
+	return inte
+
+}
+
+func histogram (s []float64) *plot.Plot {
+
+	// builds the histogram
+	// input : s = slice containing the values to plot
+
+	p, err := plot.New()
+	if err != nil {
+		log.Panic(err)
+	}
+
+	val := make(plotter.Values, len(s))
+	for i,_ := range s {
+		val[i] = s[i]
+	}
+
+	p.Title.Text = "Histogram"
+	h, err := plotter.NewHist(val, len(val))
+	if err != nil {
+		log.Panic(err)
+	}
+	p.Add(h)
+
+	return p
+}
+
+/*
+######
+Main function
+######
+*/
+
 func main() {
+
+	// reads the raw datas and format it into a slice of strings, each containing the information for a SN
+	datas, err := ioutil.ReadFile("./data/jla_lcparams.txt")
+	if err != nil {
+		fmt.Println(err)
+	}
+	str := strings.Trim(string(datas), "\n")		// gets rid, if needed, of a new line symbol at the end of the file
+	supernovae := strings.Split(str, "\n")			// builds a slice with data for a SN in each item
+	supernovae = append(supernovae[:0], supernovae[1:]...)	// suppress the first item (labels of the columns)
+
+	// creates slices with the needed values from JLA datas
+
+	sn_names := make([]string, 0)	// names
```

Comment:
you know _a priori_ the final lengths of these slices. just create them with it:

```go
var (
    N = len(supernovae)
    sn_names = make([]string, N) // names of the supernovae
    zcmb = make([]float64, N)    // redshift in the CMB frame
    ...
)
```